Custom OpenAI endpoints + Fixing HuggingFace custom endpoints

peterwilli opened the pull request:
This PR allows to set OpenAI endpoints, and also fixes the "failed to parse model name" on HuggingFace when setting custom endpoints.

I slightly refactored the way that endpoints are handled so that we dont have to extract the URL for each time we create a new outsource model.

For OpenAI, setting custom endpoints has plenty of usecases, for example, [FastChat](https://github.com/lm-sys/FastChat) has the latest state of the art to run models quickly, and has a OpenAI compatible server, using this really fits in line with the idea of running your own coding assistant.

Possible todo is to refactor the code further and use the `url` crate (already present in your application) to extract keywords after the model name rather than hard-extracting the endpoint from the model name.

This also paves the way for data points outside of the endpoint if that is ever needed!

Also, this is my first PR here, I'm open for feedback.

Also, I noticed you are the dev behind Brave! I love the browser and what you're doing!

peterwilli commented:
Updated the PR to be in sync with the current master branch

DJAndries commented on core-lib/src/generation.rs:
```diff
@@ -119,8 +120,9 @@ impl LLMVMCore {
-            ModelDescription::from_str(&model).map_err(|_| CoreError::ModelDescriptionParse)?;
+        let model_description = ModelDescription::from_str(&model)
+            .map_err(|_| CoreError::ModelDescriptionParse)
+            .unwrap();
```
Is there a reason why we unwrap here? Or is this just a left over from debugging maybe?

peterwilli replied:
Yes, I had issues with the parser ignoring any error, so I had to check that for debugging.

DJAndries commented on protocol/src/lib.rs:
```diff
@@ -206,14 +209,32 @@ impl FromStr for ModelDescription {
         }
+        let maybe_endpoint = if tokens.len() >= 3 {
+            let tail = tokens[3..].join("/");
+            // TODO: Maybe use the url crate for query string parsing? (see: https://docs.rs/url/1.4.0/url/form_urlencoded/fn.parse.html)
```
so you're thinking we would use `form_urlencoded::parse` so we can handle additional parameters, if need be? it's a good idea. happy to keep this as a todo for now :+1:

peterwilli replied:
I missed this one - Yeah something like that would work if we keep the format the same (see my comment here https://github.com/DJAndries/llmvm/pull/2#issuecomment-1801789056)

DJAndries commented on protocol/src/lib.rs:
```diff
@@ -206,14 +209,32 @@ impl FromStr for ModelDescription {
+        if tokens.len() < 3 || tokens[..2].iter().any(|v| v.is_empty()) {
             return Err(());
         }
+        let maybe_endpoint = if tokens.len() >= 3 {
```
nit: should we call this `endpoint`? perhaps we could just initialize this to `None`, and then set it to `Some(...)` within the if blocks, so we can avoid the "else ... None" blocks here:

https://github.com/DJAndries/llmvm/blob/21e50c47652a2cc3f3f1a7acde2cd74d2e37cd7d/protocol/src/lib.rs#L226-L231

peterwilli replied:
Sure, I'm ok with that

DJAndries commented on protocol/src/lib.rs:
```diff
@@ -206,14 +209,32 @@ impl FromStr for ModelDescription {
             backend: tokens_iter.next().unwrap(),
             provider: tokens_iter.next().unwrap(),
-            model_name: tokens_iter.collect::<Vec<String>>().join("/"),
+            model_name: tokens_iter.next().unwrap(),
```
i'm afraid removing the join will break any requests that include model names with slashes embedded. take the huggingface model `bigcode/starcoder` for example.

peterwilli replied:
You're right, that's true. I didn't think of that. This will most likely be solved by my proposal in https://github.com/DJAndries/llmvm/pull/2#issuecomment-1801774036

DJAndries reviewed (changes requested):
Thanks a bunch for the PR, much appreciated! This is a great improvement, especially considering the amount of community applications using OpenAI's API.

What do you think about also adding an `openai_endpoint` to the outsource config toml, so that all OpenAI requests are automatically forwarded to a pre-configured endpoint? Would that be useful to you?

I just had a few comments regarding some small things. If you're too busy to address, no worries, I can polish this up next week if need be. 

Also, glad to hear you're loving the browser! :smile:

peterwilli commented:
Hey @DJAndries thanks for responding! I'll go over it bit by bit.

> What do you think about also adding an `openai_endpoint` to the outsource config toml, so that all OpenAI requests are automatically forwarded to a pre-configured endpoint? Would that be useful to you?

My default behavior when it comes to contributing to other's projects is not breaking default behavior (So I kept the same format as with HuggingFace), I think this is a more clean solution! I think we should have something similar for HuggingFace too then (since they are both inference endpoints), this would be much easier to parse too with serde.


> I just had a few comments regarding some small things. If you're too busy to address, no worries, I can polish this up next week if need be.

It's all fine! I was working on a similar tool myself over the weekend, but you did a way better job I could have ever done in just 2 days. I'll likely add more features in the near future, so I'll be up for making changes.

peterwilli commented:
Ok I read your feedback too. I think we can either:

1. fix the things in the join "/" handling (Making sure HF doesn't break)
2. we can change the toml format to add model endpoint.

But I think that if we do the second option, we might as well do it properly, and make the format like this:

```toml
architecture = [ollama, huggingface, openai]
api_key = ...
endpoint = http://....
```

Where endpoint and/or api_key are required. Of course, this will break current behavior, so we might add a deprecation warning when old configs are found. This is the most invasive change, but it's really clean IMO, and prevents clutter where we have endpoint variables for each model i.e `ollama_endpoint`, `huggingface_endpoint`, etc

I'll let you decide what is the best way forward, I'll be happy to do either change, feel free to ping me when I can continue on this!

DJAndries commented:
> Ok I read your feedback too. I think we can either:
fix the things in the join "/" handling (Making sure HF doesn't break)
we can change the toml format to add model endpoint.

IMO, it would be cool to have both! It gives the user options.

> But I think that if we do the second option, we might as well do it properly, and make the format like this:

While I appreciate your idea of making the configuration more structured/clean, I think we should put off any breaking changes if possible. Especially considering that there will only be 7 keys in the outsource config once the new key is added. It might a good idea to restructure in the future if this config grows, so perhaps we should just leave it as a TODO for now.

peterwilli commented:
> > Ok I read your feedback too. I think we can either:
> > fix the things in the join "/" handling (Making sure HF doesn't break)
> > we can change the toml format to add model endpoint.
> 
> IMO, it would be cool to have both! It gives the user options.
> 
> > But I think that if we do the second option, we might as well do it properly, and make the format like this:
> 
> While I appreciate your idea of making the configuration more structured/clean, I think we should put off any breaking changes if possible. Especially considering that there will only be 7 keys in the outsource config once the new key is added. It might a good idea to restructure in the future if this config grows, so perhaps we should just leave it as a TODO for now.

Understandable. I will get started then, probably done in a day or 2

peterwilli commented:
Hey @DJAndries, I'm sorry for such a late response! I just finished your feedback. It took so long because I got sick and when I recovered I had a lot of work to catch up to.

I renamed the variables, fixed the endpoint parser (making sure it only takes `endpoint=` at the end of the string, so that model names like 'username/modelname` remain parsed correctly. I hope it finds you well.

Also, my colleague asked me to tell you he also enjoys the Brave browser very much, even going as far as saying that accessing the internet without it is unimaginable!

DJAndries commented:
> Hey @DJAndries, I'm sorry for such a late response! I just finished your feedback. It took so long because I got sick and when I recovered I had a lot of work to catch up to.
> 
> I renamed the variables, fixed the endpoint parser (making sure it only takes `endpoint=` at the end of the string, so that model names like 'username/modelname` remain parsed correctly. I hope it finds you well.
> 
> Also, my colleague asked me to tell you he also enjoys the Brave browser very much, even going as far as saying that accessing the internet without it is unimaginable!

Hey @peterwilli , thanks for the update! i'll give this a test and will publish a new version soon :smiley:
